sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/models/BlobDownloadAsyncResponse.java

```diff
@@ -0,0 +1,112 @@
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.function.Function;
+
+/**
+ * This class contains the response information returned from the server when downloading a blob.
+ *
+ * <p>
+ * Failed downloads are automatically retied based on the {@link ReliableDownloadOptions download options}, the retry
+ * will be resumed from the point where the download failed. This allows for the download to be consumed as one
+ * continuous stream.
+ * </p>
+ */
+public final class BlobDownloadAsyncResponse extends ResponseBase<BlobDownloadHeaders, Flux<ByteBuffer>> {
+    private final ReliableDownloadOptions options;
+    private final HttpGetterInfo info;
+    private final Function<HttpGetterInfo, Mono<BlobDownloadAsyncResponse>> getter;
+
+    /**
+     * Constructs a {@link BlobDownloadAsyncResponse}.
+     *
+     * @param request Request sent to the service.
+     * @param statusCode Response status code returned by the service.
+     * @param headers Raw headers returned in the response.
+     * @param value Stream of download data being returned by the service.
+     * @param deserializedHeaders Headers deserialized into an object.
+     * @param options {@link ReliableDownloadOptions} used to retry downloads is errors occur.
+     * @param info {@link HttpGetterInfo} used to retry a partially failed download from a specific offset.
+     * @param getter Function used to attempt to retry a partially failed download.
+     */
+    public BlobDownloadAsyncResponse(HttpRequest request, int statusCode, HttpHeaders headers, Flux<ByteBuffer> value,
+        BlobDownloadHeaders deserializedHeaders, ReliableDownloadOptions options, HttpGetterInfo info,
+        Function<HttpGetterInfo, Mono<BlobDownloadAsyncResponse>> getter) {
+        super(request, statusCode, headers, value, deserializedHeaders);
+
+        StorageImplUtils.assertNotNull("getter", getter);
+        StorageImplUtils.assertNotNull("info", info);
+        StorageImplUtils.assertNotNull("info.eTag", info.getETag());
+        this.info = info;
+        this.getter = getter;
+        this.options = (options == null) ? new ReliableDownloadOptions() : options;
+    }
+
+    @Override
+    public Flux<ByteBuffer> getValue() {
+        /*
+        We pass -1 for currentRetryCount because we want tryContinueFlux to receive a value of 0 for number of
+        retries as we have not actually retried yet, only made the initial try. Because applyReliableDownload() will
+        add 1 before calling into tryContinueFlux, we set the initial value to -1.
+         */
+        Flux<ByteBuffer> value = (options.maxRetryRequests() == 0)
+            ? super.getValue()
+            : applyReliableDownload(super.getValue(), -1, options);
+
+        return value.switchIfEmpty(Flux.just(ByteBuffer.wrap(new byte[0])));
+    }
+
+    private Flux<ByteBuffer> tryContinueFlux(Throwable t, int retryCount, ReliableDownloadOptions options) {
```

Comment:
Could this be replaced with retryWhen? The companion publisher outputting the offset that we were last at or completion if we are out of retries.